### ISSUE TITLE
We need Simple Image

### ISSUE THREAD

**Issue by mahotilo** [Typesetter/Typesetter#663]
I have tried the built-in Image section again with the new site and found it ...um... useless with responsive theme.
We do need promised Simple Image as the new default image section!!!
Especially when the new BS4 theme is coming:)

What are your plans, @juek?

**Comment by juek** [comment]
> What are your plans?

Well, it's still that “one type to rule them all, one click to find them, to crop, resize, align them, to parallax and focuspoint, to tint and link and lazyload, and in the darkness bind them”. I'm not quite sure about the last one, though, hmm 
;)

**Comment by mahotilo** [comment]
Wow. I have just dreamed about more flexible sizing:)

**Comment by juek** [comment]
Thinking process initialized …

#### Section Dimensions
○ default (defined by context)
○ custom width  [        ] ( dropdown, opts: % of parent width ** | px | rem | vw | vh )
    custom height [        ] ( dropdown, opts: % of parent height * | px | rem | vw | vh )

#### Section Aspect Ratio
○ default (defined by context)
○ 2 by 1 * or **
○ 16 by 9 * or **
○ 4 by 3 * or **
○ 1 by 1 (square) * or **
○ 3 by 4 * or **
○ custom [     ] by [     ] * or **

#### Image Size
○ natural size
○ fit horizontally **
○ fit vertically *
○ contain *
○ cover * ( set focuspoint btn )
○ custom width  [        ] ( dropdown, opts: % of section width ** | px | rem | vw | vh )
○ custom height [        ] ( dropdown, opts: % of section height * | px | rem | vw | vh )

#### Image Shape
○ original (rectangular)
○ elliptical
○ pill
○ circle
○ square

#### Image Rotation
   |—————○—————|
-180º             0º              180º

#### Image Alignment
horizontal **
   |—————○—————|
left              center           right

vertical *
   |—————○—————|
top             center          bottom

#### Link
[ _url / page-autocomplete_ ] ( select file btn )
open in 
○ same page
○ new tab
○ colorbox

#### Effects
□ parallax * ( set parallax params )
□ vignetting
□ add slight noise

#### Hover Effects
□ darken by 15%
□ lighten by 15%
□ scale to 110%

#### Caption
[_insert caption_]

#### Caption Position
○ below image
○ bottom overlay bar
○ fit-to-fill overlay

#### Caption Color
text-color [ #ffffff_]
background-color [ #222222 ]                     

More Options
□ lazy-load

---
* section needs a (min-)height defined by inline-style, custom css class or flexbox context
** section needs a (min-)width due to block-level or defined by inline-style, css (grid class, flex property, utility class)
    horz/vert alignment only works for images that are smaller than the section (for whatever reason/setting)

This UI/UX will be bit of a challenge.

I will most likely *not implement* server-side resized image creation. Results from PHP/gd are suboptimal at best, not to say 'bad'. Manual image optimization will normally lead to much better results and smaller sizes.


Any ideas what's still missing?

**Comment by mahotilo** [comment]
Wow 8O

Link
○ **[ image itself in colorbox]**
○ [ url / page-autocomplete ] ( select file btn )


I am still reading

**Comment by mahotilo** [comment]
> I will most likely not implement server-side resized image creation. Results from PHP/gd are suboptimal at best, not to say 'bad'.

Maybe, some logic should be added for resizing? 

In my experience, many inexperienced end-users tend to simply upload huge files from their smartphones. If the image is scaled from Full HD to 300px, resizing might be very useful.
We could turn on auto resizing after a certain threshold, e.g. if the natural size is 2x-3x bigger than output. And we can resize the image to a slightly bigger size than the output to mitigate imperfect PHP/gd results.

**Comment by mahotilo** [comment]
Show Caption
○ always
○ on hover

**Comment by juek** [comment]
> If the image is scaled from Full HD to 300px, resizing might be very useful.
> UPD:
> This might be the rule for TS, not just for the Image section.

>  e.g. if the natural size is 2x-3x bigger than output. 

Yes, but the output size will vary. Let's say, when a col-3 image section is edited on desktop it will show smaller than mobile-first (col-12) on a 6-inch phone screen (even css-pixel-wise, not talking about device pixels). So it's hard to estimate an optimal size.

>  … to mitigate imperfect PHP/gd results.

The problem with gd is that it cannot optimize images. I hope the following image will not be harmed too much when uploaded to GitHub, let's see....

![grafik](https://user-images.githubusercontent.com/704283/94055483-4aa2bb80-fddd-11ea-8751-135eb8eb3a92.png)

**Comment by juek** [comment]
maybe

Image Size
○ use as background of the current wrapper section

but this only makes sense with certain settings. e.g. 'cover' and/or 'parallax'

**Comment by mahotilo** [comment]
Yep. Col-x sizing is a problem.

**Comment by mahotilo** [comment]
> ○ use as background of the current wrapper section

I'm afraid to imagine what the gallery might become after the similar tuning. The Slider factory :D

**Comment by juek** [comment]
and of course

Section Display

○ default (block/flex, defined by context)
○ inline

the latter is ideal if you e.g. want to build a row of logos with different aspect ratios, best inside a flex context

**Comment by mahotilo** [comment]
Will paralax/focuspoint  related JS create a permanent footprint on a speed of simple page with plain images?

**Comment by juek** [comment]
> … gallery might become after the similar tuning

Gallery will definitely continue to use gd thumbs but it's css needs to be 'flexboxed'. Will spare us some 20 KB css at minimum.

> Slider factory ?

if everything goes right, which is far from certain, I'm planning to convert 'Responsive Image', 'Simple Image' and 'Parallax Image' to become the new image type, once they're edited.

> Will paralax/focuspoint related JS create a permanent footprint on a speed of simple page with plain images?

Yes and No.
Yes because it will be all the same JS that needs to be loaded. 
And No because focuspoint and parallax math will only be done on images using these features.

**Comment by mahotilo** [comment]
> I'm planning to convert 'Responsive Image', 'Simple Image' and 'Parallax Image' to become the new image type, once they're edited.

Do you mean, I inserted a section of one type, and after changing the option it can became a section of another type?

**Comment by juek** [comment]
> UPD
> Or, will the assets be loaded depending on the selected features?

Runtime JS footprint for parallax and focuspoint is neglible, alltogether ~7 KB (minified)

**Comment by juek** [comment]
> Do you mean, I inserted a section of one type, and after changing the option it can become a section of another type?

If it was a Reponsive Image section before and you edit it, it will become an AllMightyImage section. From then you can switch/mix the behavior on the fly. Same applies to the other 2 types. 
I'd also like to convert the current default images but this will be harder to do. 

But only If everything goes right. 
I haven't done type conversion before but it should be possible.

**Comment by mahotilo** [comment]
> it will become an AllMightyImage section

This should be a TS option, do not you think?

**Comment by juek** [comment]
> From then you can switch/mix the behavior on the fly.

The problem is not mixing the functionality. It's the editor UI that must communicate meaningful combinations.
Basically there is no reason not to parallax even an inline Simple-Image-like set AllMightyImage. But it's hard to get the settings right.

> This should be a TS settings option, do not you think?

If I actually manage to do a 1:1 full conversion (same behavior) I can't see a reason not to do it. I'd like to make AllMightyImage the default image type in TS.

**Comment by juek** [comment]
I remember when I initially made Parallax Image, folks asked why they cannot set more parameters. My answer then was: 
Trust me, you do not want to :)

**Comment by mahotilo** [comment]
But, we still have this
![image](https://user-images.githubusercontent.com/14929385/94060429-a294ee80-fdec-11ea-9c0c-96c3e6e0c273.png)

Are we getting rid of this too?

**Comment by juek** [comment]
> Are we getting rid of this too?

Maybe. With some good flex based defaults, we probably could, because they would be much easier to overrule. The problem here is that older sites may still overrule the legacy style with their custom one. Some of mine do that. These custom rulesets will definitely not be able to completely overrule the modern style without some additions.

**Comment by mahotilo** [comment]
> The problem here is that older sites may still overrule the legacy style with their custom one. 

Could this be a problem with Simple Image auto conversion to Mighty one? But to be honest, I barely used Image section in real projects.

**Comment by juek** [comment]
> I'd like to make AllMightyImage the default image type in TS.

Emphasis on 'I'd like to'. If it turns out to be too complicated/laborious/buggy, it will become a plugin first and TS 5.2 will go without it. It's really time for get 5.2 out.

It's only that the new BS4 theme could greatly benefit from that image type. I could add a few nice Section Combos, like 'Card Row' / 'Card Deck', 'Testimonials' (with images in circles), 'About Me', and such. 

![grafik](https://user-images.githubusercontent.com/704283/94062102-8c842f80-fde6-11ea-8bb5-ee84789762a5.png)

**Comment by mahotilo** [comment]
> ...new BS4 theme could greatly benefit from that image type.

True!

**Comment by juek** [comment]
> Could this be a problem with Simple Image auto conversion to AllMighty one?

Yes it could. Especially if one made special CSS rules for these image types. While the goal is to make the output look exactly the same, the markup will certainly differ.

Maybe it will rather become a 'try conversion' button. 
OMG, what bold claims I am making here?

**Comment by mahotilo** [comment]
After customizer.php almost anything is possible ...

**Comment by juek** [comment]
[List](https://github.com/Typesetter/Typesetter/issues/663#issuecomment-697715016) updated

**Comment by mahotilo** [comment]
Backface-visibility Caption?

**Comment by juek** [comment]
> Backface-visibility Caption?

Added it as a suboption on show --> on hover/touch

**Comment by juek** [comment]
The rotate + crop dilemma:

![grafik](https://user-images.githubusercontent.com/704283/94199126-a0986180-feb8-11ea-8619-da82e6ba36b5.png)

IMO this problem cannot be automatically resolved easily. So the question is if rotation should actually be a section editor feature.

**Comment by juek** [comment]
... but even for the first, 'easy' case there is not a single 'correct' solution:

![grafik](https://user-images.githubusercontent.com/704283/94199938-e275d780-feb9-11ea-9511-efb0d5ab9d03.png)

**Comment by juek** [comment]
So, if we want to keep our workflow non-destructive, which I would greatly prefer we need …

* some kind of 'Rotate+Crop workbench' editor mode to set the cropping manually
* to detect possible changes to the source image OR an adaptive focuspoint algorithm

**Comment by juek** [comment]
Added to the feature list:  Image background color (optional, for images with transparency)

**Comment by mahotilo** [comment]
We already have this possibility to edit an image rotation|crop
![image](https://user-images.githubusercontent.com/14929385/94228861-a452d500-ff06-11ea-9905-7d597b764dec.png)

Thus I have thought of rotation in the Image Section this way
![image](https://user-images.githubusercontent.com/14929385/94228799-7a99ae00-ff06-11ea-81ab-71808a413484.png)

I have thought of it as another funny effect.

**Comment by juek** [comment]
Scale, Crop and Rotate will be an interactive operation  I have already found a way to do it. We will be able to create all variants

Example:

![grafik](https://user-images.githubusercontent.com/704283/94277461-e2bdb380-ff49-11ea-8560-f04d9acef53b.png)

**Comment by juek** [comment]
This will be (most likely) the markup and functional structure:

![grafik](https://user-images.githubusercontent.com/704283/94286762-b871f300-ff55-11ea-84a3-a5fe4d578c64.png)

**Comment by mahotilo** [comment]
> use as background for the current wrapper section *

So, would this option be for div.filetype_picture or for wraper section?

**Comment by juek** [comment]
> So, would this option be for div.filetype_picture or for wraper section?

For the div.filetype-picture section itself. It however requires position:relative on the wrapper section, which cannot be applied from out of the editor. So it will rather be a preset used in a special Section Combo. Similar to Responsive Background or Parallax Background.
